File: ExcelApi/Services/AuthService.cs
using ExcelApi.DTOs;
using ExcelApi.Models;

namespace ExcelApi.Services
{
    public interface IAuthService
    {
        Task<LoginResponse> Login(LoginRequest request);
        Task<RefreshTokenResponse> RefreshToken(RefreshTokenRequest request);
        Task<UserDto> Register(RegisterRequest request);
        Task Logout(int userId);
    }

    public class AuthService : IAuthService
    {
        private readonly ExcelApi.Data.ExcelContext _context;
        private readonly IConfiguration _configuration;
        private readonly ILogger<AuthService> _logger;

        public AuthService(ExcelApi.Data.ExcelContext context, IConfiguration configuration, ILogger<AuthService> logger)
        {
            _context = context;
            _configuration = configuration;
            _logger = logger;
        }

        public async Task<LoginResponse> Login(LoginRequest request)
        {
            var user = await _context.Users.FirstOrDefaultAsync(u => u.Username == request.Username);
            if (user == null || !BCrypt.Net.BCrypt.Verify(request.Password, user.PasswordHash))
            {
                _logger.LogWarning($"Failed login attempt for user: {request.Username}");
                throw new UnauthorizedAccessException("Invalid username or password");
            }

            var token = GenerateJwtToken(user);
            var refreshToken = GenerateRefreshToken(user.Id);

            return new LoginResponse
            {
                AccessToken = token,
                RefreshToken = refreshToken.Token,
                ExpiresIn = 3600,
                User = MapToUserDto(user)
            };
        }

        public async Task<RefreshTokenResponse> RefreshToken(RefreshTokenRequest request)
        {
            var refreshToken = await _context.RefreshTokens
                .FirstOrDefaultAsync(rt => rt.Token == request.RefreshToken && !rt.IsRevoked);

            if (refreshToken == null || refreshToken.ExpiryDate < DateTime.UtcNow)
            {
                throw new UnauthorizedAccessException("Invalid or expired refresh token");
            }

            var user = await _context.Users.FindAsync(refreshToken.UserId);
            if (user == null)
            {
                throw new UnauthorizedAccessException("User not found");
            }

            var newToken = GenerateJwtToken(user);

            return new RefreshTokenResponse
            {
                AccessToken = newToken,
                ExpiresIn = 3600
            };
        }

        public async Task<UserDto> Register(RegisterRequest request)
        {
            if (await _context.Users.AnyAsync(u => u.Username == request.Username))
            {
                throw new InvalidOperationException("Username already exists");
            }

            if (await _context.Users.AnyAsync(u => u.Email == request.Email))
            {
                throw new InvalidOperationException("Email already exists");
            }

            var user = new User
            {
                Username = request.Username,
                Email = request.Email,
                FirstName = request.FirstName,
                LastName = request.LastName,
                PasswordHash = BCrypt.Net.BCrypt.HashPassword(request.Password),
                IsActive = true
            };

            _context.Users.Add(user);
            await _context.SaveChangesAsync();

            _logger.LogInformation($"New user registered: {user.Username}");

            return MapToUserDto(user);
        }

        public async Task Logout(int userId)
        {
            var refreshTokens = await _context.RefreshTokens
                .Where(rt => rt.UserId == userId && !rt.IsRevoked)
                .ToListAsync();

            foreach (var token in refreshTokens)
            {
                token.IsRevoked = true;
            }

            await _context.SaveChangesAsync();
            _logger.LogInformation($"User {userId} logged out");
        }

        private string GenerateJwtToken(User user)
        {
            var tokenHandler = new System.IdentityModel.Tokens.Jwt.JwtSecurityTokenHandler();
            var key = System.Text.Encoding.ASCII.GetBytes(_configuration["JwtSettings:Secret"] ?? "your-secret-key-here");

            var tokenDescriptor = new Microsoft.IdentityModel.Tokens.SecurityTokenDescriptor
            {
                Subject = new System.Security.Claims.ClaimsIdentity(new[]
                {
                    new System.Security.Claims.Claim("id", user.Id.ToString()),
                    new System.Security.Claims.Claim(System.Security.Claims.ClaimTypes.NameIdentifier, user.Username),
                    new System.Security.Claims.Claim(System.Security.Claims.ClaimTypes.Email, user.Email)
                }),
                Expires = DateTime.UtcNow.AddHours(1),
                SigningCredentials = new Microsoft.IdentityModel.Tokens.SigningCredentials(new Microsoft.IdentityModel.Tokens.SymmetricSecurityKey(key), Microsoft.IdentityModel.Tokens.SecurityAlgorithms.HmacSha256Signature)
            };

            var token = tokenHandler.CreateToken(tokenDescriptor);
            return tokenHandler.WriteToken(token);
        }

        private RefreshToken GenerateRefreshToken(int userId)
        {
            var refreshToken = new RefreshToken
            {
                UserId = userId,
                Token = System.Guid.NewGuid().ToString("N"),
                ExpiryDate = DateTime.UtcNow.AddDays(7),
                IsRevoked = false
            };

            _context.RefreshTokens.Add(refreshToken);
            _context.SaveChanges();

            return refreshToken;
        }

        private UserDto MapToUserDto(User user)
        {
            return new UserDto
            {
                Id = user.Id,
                Username = user.Username,
                Email = user.Email,
                FirstName = user.FirstName,
                LastName = user.LastName,
                CreatedAt = user.CreatedAt
            };
        }
    }
}

File: ExcelAddin/src/services/AuthService.ts
import axios from 'axios'

interface AuthToken {
  accessToken: string
  refreshToken: string
  expiresIn: number
}

interface LoginRequest {
  username: string
  password: string
}

interface LoginResponse {
  accessToken: string
  refreshToken: string
  expiresIn: number
}

export class AuthService {
  private static readonly API_URL = localStorage.getItem('apiUrl') || 'https://localhost:7001'

  static async login(credentials: LoginRequest): Promise<AuthToken> {
    try {
      const response = await axios.post<LoginResponse>(`${this.API_URL}/api/auth/login`, credentials)
      const { accessToken, refreshToken, expiresIn } = response.data

      localStorage.setItem('authToken', accessToken)
      localStorage.setItem('refreshToken', refreshToken)
      localStorage.setItem('tokenExpiry', (Date.now() + expiresIn * 1000).toString())

      return { accessToken, refreshToken, expiresIn }
    } catch (error) {
      throw new Error('Login failed')
    }
  }

  static async logout(): Promise<void> {
    localStorage.removeItem('authToken')
    localStorage.removeItem('refreshToken')
    localStorage.removeItem('tokenExpiry')
  }

  static async refreshToken(): Promise<string> {
    const refreshToken = localStorage.getItem('refreshToken')
    if (!refreshToken) {
      throw new Error('No refresh token available')
    }

    try {
      const response = await axios.post<{ accessToken: string; expiresIn: number }>(
        `${this.API_URL}/api/auth/refresh`,
        { refreshToken },
      )
      const { accessToken, expiresIn } = response.data

      localStorage.setItem('authToken', accessToken)
      localStorage.setItem('tokenExpiry', (Date.now() + expiresIn * 1000).toString())

      return accessToken
    } catch (error) {
      throw new Error('Token refresh failed')
    }
  }

  static isTokenExpired(token: string): boolean {
    try {
      const expiry = localStorage.getItem('tokenExpiry')
      return !expiry || Date.now() > parseInt(expiry, 10)
    } catch {
      return true
    }
  }

  static getToken(): string | null {
    const token = localStorage.getItem('authToken')
    if (token && this.isTokenExpired(token)) {
      return null
    }
    return token
  }
}
